src/Neo/SmartContract/Native/ContractManagement.cs

```diff
@@ -143,6 +146,39 @@ public ContractState GetContract(DataCache snapshot, UInt160 hash)
             return snapshot.TryGet(CreateStorageKey(Prefix_Contract).Add(hash))?.GetInteroperable<ContractState>();
         }
 
+        /// <summary>
+        /// Maps specified ID to deployed contract.
+        /// </summary>
+        /// <param name="snapshot">The snapshot used to read data.</param>
+        /// <param name="id">Contract ID.</param>
+        /// <returns>The deployed contract.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        public ContractState GetContractById(DataCache snapshot, int id)
+        {
+            StorageItem item = snapshot.TryGet(CreateStorageKey(Prefix_ContractHash).AddBigEndian(id));
+            if (item is null) return null;
+            var hash = new UInt160(item.Value.Span);
+            return GetContract(snapshot, hash);
+        }
+
+        /// <summary>
+        /// Gets hashes of all deployed contracts.
+        /// </summary>
+        /// <param name="snapshot">The snapshot used to read data.</param>
+        /// <returns>Iterator with hashes of all deployed contracts.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.ReadStates)]
+        private IIterator GetContractHashes(DataCache snapshot)
+        {
+            const FindOptions options = FindOptions.RemovePrefix;
+            byte[] prefix_key = CreateStorageKey(Prefix_ContractHash).ToArray();
+            var enumerator = snapshot.Find(prefix_key)
+                .Select(p => (p.Key, p.Value, Id: BinaryPrimitives.ReadInt32BigEndian(p.Key.Key.Span[1..])))
+                .Where(p => p.Id >= 0)
```

Comment:
We don't return native contracts?

Reply:
Yes, that's intentional. They're well-known (contract-side every devpack has their hashes hardcoded, RPC-wise we have `getnativecontracts`) and they have negative IDs which means additional code is needed to return them in order, not worth it.